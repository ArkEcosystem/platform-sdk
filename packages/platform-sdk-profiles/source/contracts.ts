export * from "./authenticator.contract";
export * from "./cache.contract";
export * from "./coin.service.contract";
export * from "./contact-address.repository.contract";
export * from "./contact-address.contract";
export * from "./contact.repository.contract";
export * from "./contact.contract";
export * from "./count.aggregate.contract";
export * from "./data.repository.contract";
export * from "./delegate.service.contract";
export * from "./exchange-rate.service.contract";
export * from "./fee.service.contract";
export * from "./known-wallet.service.contract";
export * from "./migrator.contract";
export * from "./multi-signature.contract";
export * from "./notification.repository.contract";
export * from "./password.contract";
export * from "./peer.repository.contract";
export * from "./plugin-registry.contract";
export * from "./plugin.repository.contract";
export * from "./portfolio.contract";
export * from "./profile.repository.contract";
export * from "./profile.contract";
export * from "./profile.dumper.contract";
export * from "./profile.encrypter.contract";
export * from "./profile.enum.contract";
export * from "./profile.exporter.contract";
export * from "./profile.factory.contract";
export * from "./profile.importer.contract";
export * from "./profile.initialiser.contract";
export * from "./profile.serialiser.contract";
export * from "./profile.status.contract";
export * from "./profile.validator.contract";
export * from "./read-only-wallet.contract";
export * from "./registration.aggregate.contract";
export * from "./setting.repository.contract";
export * from "./transaction.aggregate.contract";
export * from "./transaction-index.contract";
export * from "./vote-registry.contract";
export * from "./wallet.aggregate.contract";
export * from "./wallet.repository.contract";
export * from "./wallet.service.contract";
export * from "./wallet-transaction.service.contract";
export * from "./wallet.contract";
export * from "./wallet.enum"; // @TODO
export * from "./wallet.factory.contract";
export * from "./wallet.gate.contract";
export * from "./wallet.mutator.contract";
export * from "./wallet.synchroniser.contract";
export * from "./wif.contract";
