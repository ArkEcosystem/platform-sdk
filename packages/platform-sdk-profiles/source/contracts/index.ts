export * from "./contacts";
export * from "./driver";
export * from "./plugins";
export * from "./profiles";
export * from "./repositories";
export * from "./services";
export * from "./wallets";
