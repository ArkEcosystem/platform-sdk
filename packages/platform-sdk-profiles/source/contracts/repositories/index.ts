export * from "./data-repository";
export * from "./contact-address-repository";
export * from "./contact-repository";
export * from "./notification-repository";
export * from "./peer-repository";
export * from "./profile-repository";
export * from "./setting-repository";
export * from "./wallet-repository";
