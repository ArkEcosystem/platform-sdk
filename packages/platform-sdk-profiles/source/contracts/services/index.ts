export * from "./coin-service";
export * from "./cache";
export * from "./delegate-service";
export * from "./exchange-rate-service";
export * from "./fee-service";
export * from "./known-wallet-service";
export * from "./wallet-service";
