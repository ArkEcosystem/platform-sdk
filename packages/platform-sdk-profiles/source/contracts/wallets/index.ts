export * from "./read-only-wallet";
export * from "./wallet-transaction-service";
export * from "./wallet.enum";
export * from "./wallet.factory";
export * from "./wallet";
