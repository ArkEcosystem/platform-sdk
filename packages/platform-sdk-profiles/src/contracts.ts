export interface EnvironmentOptions {
	storage: string;
	migrations?: Record<string, any>;
}

export interface Storage {
	all(): Promise<object>;

	get<T>(key: string): Promise<T | undefined>;

	set(key: string, value: string | object): Promise<void>;

	forget(key: string): Promise<void>;

	flush(): Promise<void>;

	count(): Promise<number>;

	snapshot(): Promise<void>;

	restore(): Promise<void>;
}
