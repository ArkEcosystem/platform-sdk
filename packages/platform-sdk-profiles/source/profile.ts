/* eslint-disable @typescript-eslint/no-non-null-assertion */
import { BigNumber } from "@arkecosystem/platform-sdk-support";
import {
	IContactRepository,
	IPortfolio,
	ICountAggregate,
	IDataRepository,
	INotificationRepository,
	IPeerRepository,
	IPluginRepository,
	IProfile,
	IProfileInput,
	IRegistrationAggregate,
	ISettingRepository,
	ITransactionAggregate,
	IWalletAggregate,
	IWalletRepository,
	ProfileSetting,
	IReadWriteWallet,
	ICoinService,
	IAuthenticator,
	IWalletFactory,
	ProfileData,
	IProfileStatus,
} from "./contracts";

import { PluginRepository } from "./plugin.repository";
import { ContactRepository } from "./contact.repository";
import { DataRepository } from "./data.repository";
import { NotificationRepository } from "./notification.repository";
import { PeerRepository } from "./peer.repository";
import { SettingRepository } from "./setting.repository";
import { WalletRepository } from "./wallet.repository";
import { Avatar } from "./helpers";
import { CountAggregate } from "./count.aggregate";
import { RegistrationAggregate } from "./registration.aggregate";
import { TransactionAggregate } from "./transaction.aggregate";
import { WalletAggregate } from "./wallet.aggregate";
import { Authenticator } from "./authenticator";
import { Portfolio } from "./portfolio";
import { CoinService } from "./coin.service";
import { WalletFactory } from "./wallet.factory";
import { AttributeBag } from "./helpers/attribute-bag";
import { ProfileInitialiser } from "./profile.initialiser";
import { IPasswordManager } from "./contracts";
import { PasswordManager } from "./password";
import { ProfileStatus } from "./profile.status";

export class Profile implements IProfile {
	/**
	 * The coin service.
	 *
	 * @type {ICoinService}
	 * @memberof Profile
	 */
	readonly #coinService: ICoinService;

	/**
	 * The portfolio service.
	 *
	 * @type {IPortfolio}
	 * @memberof Profile
	 */
	readonly #portfolio: IPortfolio;

	/**
	 * The contact repository.
	 *
	 * @type {IContactRepository}
	 * @memberof Profile
	 */
	readonly #contactRepository: IContactRepository;

	/**
	 * The data repository.
	 *
	 * @type {IDataRepository}
	 * @memberof Profile
	 */
	readonly #dataRepository: IDataRepository;

	/**
	 * The notification repository.
	 *
	 * @type {INotificationRepository}
	 * @memberof Profile
	 */
	readonly #notificationRepository: INotificationRepository;

	/**
	 * The peer repository.
	 *
	 * @type {IPeerRepository}
	 * @memberof Profile
	 */
	readonly #peerRepository: IPeerRepository;

	/**
	 * The plugin repository.
	 *
	 * @type {IPluginRepository}
	 * @memberof Profile
	 */
	readonly #pluginRepository: IPluginRepository;

	/**
	 * The setting repository.
	 *
	 * @type {ISettingRepository}
	 * @memberof Profile
	 */
	readonly #settingRepository: ISettingRepository;

	/**
	 * The wallet repository.
	 *
	 * @type {IWalletRepository}
	 * @memberof Profile
	 */
	readonly #walletRepository: IWalletRepository;

	/**
	 * The count aggregate service.
	 *
	 * @type {ICountAggregate}
	 * @memberof Profile
	 */
	readonly #countAggregate: ICountAggregate;

	/**
	 * The registration aggregate service.
	 *
	 * @type {IRegistrationAggregate}
	 * @memberof Profile
	 */
	readonly #registrationAggregate: IRegistrationAggregate;

	/**
	 * The transaction aggregate service.
	 *
	 * @type {ITransactionAggregate}
	 * @memberof Profile
	 */
	readonly #transactionAggregate: ITransactionAggregate;

	/**
	 * The wallet aggregate service.
	 *
	 * @type {IWalletAggregate}
	 * @memberof Profile
	 */
	readonly #walletAggregate: IWalletAggregate;

	/**
	 * The authentication service.
	 *
	 * @type {IAuthenticator}
	 * @memberof Profile
	 */
	readonly #authenticator: IAuthenticator;

	/**
	 * The password service.
	 *
	 * @type {IAuthenticator}
	 * @memberof Profile
	 */
	readonly #password: IPasswordManager;

	/**
	 * The normalise profile data.
	 *
	 * @type {AttributeBag<IProfileInput>}
	 * @memberof Profile
	 */
	readonly #attributes: AttributeBag<IProfileInput>;

	/**
	 * The status service.
	 *
	 * @type {IProfileStatus}
	 * @memberof Profile
	 */
	readonly #status: IProfileStatus;

	public constructor(data: IProfileInput) {
		this.#attributes = new AttributeBag<IProfileInput>(data);
		this.#coinService = new CoinService(new DataRepository());
		this.#portfolio = new Portfolio(this);
		this.#contactRepository = new ContactRepository(this);
		this.#dataRepository = new DataRepository();
		this.#notificationRepository = new NotificationRepository(this);
		this.#peerRepository = new PeerRepository(this);
		this.#pluginRepository = new PluginRepository();
		this.#settingRepository = new SettingRepository(this, Object.values(ProfileSetting));
		this.#walletRepository = new WalletRepository(this);
		this.#countAggregate = new CountAggregate(this);
		this.#registrationAggregate = new RegistrationAggregate(this);
		this.#transactionAggregate = new TransactionAggregate(this);
		this.#walletAggregate = new WalletAggregate(this);
		this.#authenticator = new Authenticator(this);
		this.#password = new PasswordManager();
		this.#status = new ProfileStatus();
	}

	/** {@inheritDoc IProfile.id} */
	public id(): string {
		return this.#attributes.get<string>("id");
	}

	/** {@inheritDoc IProfile.name} */
	public name(): string {
		if (this.settings().missing(ProfileSetting.Name)) {
			return this.#attributes.get<string>("name");
		}

		return this.settings().get<string>(ProfileSetting.Name)!;
	}

	/** {@inheritDoc IProfile.avatar} */
	public avatar(): string {
		const avatarFromSettings: string | undefined = this.settings().get(ProfileSetting.Avatar);

		if (avatarFromSettings) {
			return avatarFromSettings;
		}

		if (this.#attributes.hasStrict("data.avatar")) {
			return this.#attributes.get<string>("avatar");
		}

		return Avatar.make(this.name());
	}

	/** {@inheritDoc IProfile.theme} */
	public theme(): string {
		if (this.settings().missing(ProfileSetting.Theme)) {
			return this.#attributes.get<string>("theme") || "light";
		}

		return this.settings().get<string>(ProfileSetting.Theme)!;
	}

	/** {@inheritDoc IProfile.balance} */
	public balance(): number {
		return this.walletAggregate().balance();
	}

	/** {@inheritDoc IProfile.convertedBalance} */
	public convertedBalance(): number {
		return this.walletAggregate().convertedBalance();
	}

	/** {@inheritDoc IProfile.flush} */
	public flush(): void {
		const name: string | undefined = this.settings().get(ProfileSetting.Name);

		if (name === undefined) {
			throw new Error("The name of the profile could not be found. This looks like a bug.");
		}

		new ProfileInitialiser(this).initialise(name);
	}

	/** {@inheritDoc IProfile.initialiseSettings} */
	public flushSettings(): void {
		const name: string | undefined = this.settings().get(ProfileSetting.Name);

		if (name === undefined) {
			throw new Error("The name of the profile could not be found. This looks like a bug.");
		}

		new ProfileInitialiser(this).initialiseSettings(name);
	}

	/** {@inheritDoc IProfile.coins} */
	public coins(): ICoinService {
		return this.#coinService;
	}

	/** {@inheritDoc IProfile.portfolio} */
	public portfolio(): IPortfolio {
		return this.#portfolio;
	}

	/** {@inheritDoc IProfile.contacts} */
	public contacts(): IContactRepository {
		return this.#contactRepository;
	}

	/** {@inheritDoc IProfile.data} */
	public data(): IDataRepository {
		return this.#dataRepository;
	}

	/** {@inheritDoc IProfile.notifications} */
	public notifications(): INotificationRepository {
		return this.#notificationRepository;
	}

	/** {@inheritDoc IProfile.peers} */
	public peers(): IPeerRepository {
		return this.#peerRepository;
	}

	/** {@inheritDoc IProfile.plugins} */
	public plugins(): IPluginRepository {
		return this.#pluginRepository;
	}

	/** {@inheritDoc IProfile.settings} */
	public settings(): ISettingRepository {
		return this.#settingRepository;
	}

	/** {@inheritDoc IProfile.wallets} */
	public wallets(): IWalletRepository {
		return this.#walletRepository;
	}

	/** {@inheritDoc IProfile.walletFactory} */
	public walletFactory(): IWalletFactory {
		return new WalletFactory(this);
	}

	/** {@inheritDoc IProfile.countAggregate} */
	public countAggregate(): ICountAggregate {
		return this.#countAggregate;
	}

	/** {@inheritDoc IProfile.registrationAggregate} */
	public registrationAggregate(): IRegistrationAggregate {
		return this.#registrationAggregate;
	}

	/** {@inheritDoc IProfile.transactionAggregate} */
	public transactionAggregate(): ITransactionAggregate {
		return this.#transactionAggregate;
	}

	/** {@inheritDoc IProfile.walletAggregate} */
	public walletAggregate(): IWalletAggregate {
		return this.#walletAggregate;
	}

	/** {@inheritDoc IProfile.auth} */
	public auth(): IAuthenticator {
		return this.#authenticator;
	}

	/** {@inheritDoc IProfile.password} */
	public password(): IPasswordManager {
		return this.#password;
	}

	/** {@inheritDoc IProfile.status} */
	public status(): IProfileStatus {
		return this.#status;
	}

	/** {@inheritDoc IProfile.usesPassword} */
	public usesPassword(): boolean {
		return this.#attributes.hasStrict("password");
	}

	/** {@inheritDoc IProfile.hasBeenPartiallyRestored} */
	public hasBeenPartiallyRestored(): boolean {
		return (
			this.#walletRepository.values().filter((wallet: IReadWriteWallet) => wallet.hasBeenPartiallyRestored())
				.length > 0
		);
	}

	/** {@inheritDoc IProfile.getAttributes} */
	public getAttributes(): AttributeBag<IProfileInput> {
		return this.#attributes;
	}

	/** {@inheritDoc IProfile.async} */
	public async sync(): Promise<void> {
		await this.wallets().restore();

		await this.contacts().restore();
	}

	/** {@inheritDoc IProfile.markIntroductoryTutorialAsComplete} */
	public markIntroductoryTutorialAsComplete(): void {
		this.data().set(ProfileData.HasCompletedIntroductoryTutorial, true);

		this.status().markAsDirty();
	}

	/** {@inheritDoc IProfile.hasCompletedIntroductoryTutorial} */
	public hasCompletedIntroductoryTutorial(): boolean {
		return this.data().has(ProfileData.HasCompletedIntroductoryTutorial);
	}

	/** {@inheritDoc IProfile.markManualInstallationDisclaimerAsAccepted} */
	public markManualInstallationDisclaimerAsAccepted(): void {
		this.data().set(ProfileData.HasAcceptedManualInstallationDisclaimer, true);

		this.status().markAsDirty();
	}

	/** {@inheritDoc IProfile.hasAcceptedManualInstallationDisclaimer} */
	public hasAcceptedManualInstallationDisclaimer(): boolean {
		return this.data().has(ProfileData.HasAcceptedManualInstallationDisclaimer);
	}
}
