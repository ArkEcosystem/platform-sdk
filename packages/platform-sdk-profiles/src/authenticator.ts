import { Bcrypt } from "@arkecosystem/platform-sdk-crypto";

import { ProfileSetting } from "./profile.models";

export class Authenticator {
	readonly #profile;

	public constructor(profile) {
		this.#profile = profile;
	}

	public setPassword(password: string): void {
		this.#profile.settings().set(ProfileSetting.Password, Bcrypt.hash(password));
	}

	public verifyPassword(password: string): boolean {
		return Bcrypt.verify(this.#profile.settings().get(ProfileSetting.Password), password);
	}

	public changePassword(oldPassword: string, newPassword: string): void {
		const currentPassword: string | undefined = this.#profile.settings().get(ProfileSetting.Password);

		if (!currentPassword) {
			throw new Error("No password is set. Call [setPassword] instead.");
		}

		if (!Bcrypt.verify(currentPassword, oldPassword)) {
			throw new Error("The current password does not match.");
		}

		this.setPassword(newPassword);
	}
}
