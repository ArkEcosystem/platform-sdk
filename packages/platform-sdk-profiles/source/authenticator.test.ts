import "jest-extended";
import "reflect-metadata";

import { bootContainer } from "../test/mocking";
import { IProfile, ProfileSetting } from "./contracts";
import { Authenticator } from "./authenticator";
import { Profile } from "./profile";
import { ProfileExporter } from "./profile.exporter";

let subject: Authenticator;
let profile: IProfile;

beforeAll(() => bootContainer());

beforeEach(() => {
	profile = new Profile({ id: "uuid", name: "name", avatar: "avatar", data: "" });
	subject = new Authenticator(profile);
});

it("should set the password", async () => {
	expect(profile.settings().get(ProfileSetting.Password)).toBeUndefined();

	expect(subject.setPassword("password")).toBeUndefined();

	expect(profile.settings().get(ProfileSetting.Password)).toBeString();
});

it("should verify the password", async () => {
	subject.setPassword("password");

	expect(subject.verifyPassword("password")).toBeTrue();
	expect(subject.verifyPassword("invalid")).toBeFalse();
});

it("should fail to verify the password for a profile that doesn't use a profile", async () => {
	expect(() => subject.verifyPassword("password")).toThrow("No password is set.");
});

it("should change the password", () => {
	subject.setPassword("old-password");

	const oldPassword = profile.settings().get(ProfileSetting.Password);

	expect(subject.changePassword("old-password", "new-password")).toBeUndefined();

	expect(profile.settings().get(ProfileSetting.Password)).not.toBe(oldPassword);
});

it("should fail to change the password if no password is set", () => {
	expect(() => subject.changePassword("old-password", "new-password")).toThrow("No password");
});

it("should fail to change the password if the old password is invalid", () => {
	subject.setPassword("old-password");

	expect(() => subject.changePassword("invalid-old-password", "new-password")).toThrow("does not match");
});

it("should set password in memory", () => {
	subject.setPassword("password");

	expect(profile.password().get()).toEqual("password");
});

it("should forget the password", () => {
	expect(profile.usesPassword()).toBeFalse();
	const firstExport = new ProfileExporter(profile).export();
	expect(firstExport).toMatchSnapshot();

	subject.setPassword("old-password");

	expect(profile.usesPassword()).toBeTrue();
	expect(new ProfileExporter(profile).export().length > firstExport.length * 2).toBeTrue();

	subject.forgetPassword("old-password");

	expect(profile.usesPassword()).toBeFalse();
	expect(new ProfileExporter(profile).export().length <= firstExport.length).toBeTrue();
});
