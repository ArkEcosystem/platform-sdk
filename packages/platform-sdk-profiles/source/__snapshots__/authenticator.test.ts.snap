// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should forget the password 1`] = `"eyJpZCI6InV1aWQiLCJjb250YWN0cyI6e30sImRhdGEiOnt9LCJub3RpZmljYXRpb25zIjp7fSwicGVlcnMiOnt9LCJwbHVnaW5zIjp7fSwic2V0dGluZ3MiOnt9LCJ3YWxsZXRzIjp7fX0="`;

exports[`should forget the password 2`] = `"ZmM1ZGY1YjQwNDE1MWZiYTYxNGEyYTdlNTVjNzkxYmQ6NTE1NDM1NmQ3MzZlNTYyYjcxNjMyYjU5MzY0ZjJiNmYzMTcwNjc3NTc3NGY0ZTc4NTgzMjJmNDg2NjUzNzE0ZDMxMmYzOTQ3NDg3NzcxNGQ1NjQ2NTQ3MDRmNGM3MTU5NmMzMjQyNjQ3ODMxNmQzNDZiMzU3MzZiNGE2NzZlNDQ1OTc4NTczMDQxNTc2MTM1NTg2ODQ1MzY2NzZhNDk0NzUyNTM0ZTZjNzE2ZDU1MzY2NjUyNTczODc5NzY1MzU2NGM2ZjM2NTY0NjMxNTU0ZDQyN2E2NDdhNTA0MzM3NDc0YzM4NDE1MzM5NzY3NTRjNzI2Mjc0NWE0Yzc3NmQ0YjZmNjg3MTQxNTI0NjQ2NzY1OTRiNTM0YzMwNDE2ODM2NTkzODM2NjYzMDdhNGM1NzM0MzE3YTcxNTk0MzMwNGM3Mzc4NzczNTM2Nzk0YjRiNGM0YjRmNjg0NTc5MzI2YTdhNTA2ZjQ5NDk2NDRjNTY3NTY5NjY0YTQxNDg3MTUwNDk2ODZmNzEzNzQyNjg2MTJmNTk3MjM4NjE0MTY3NGE1MTY0NmUzMjMzNmQyYjY2NDMzNTU5NDMzMDQ5MzA3MjM4NjI0NjQ5NTI0ODZhNGQ1NTU3NGUzOTc5MmY2ZTM3NDE1Njc0NGM0MjM2NjU3ODQ3NTM2ZjJiMzA1ODY4Mzc0ZTM2NDU1NTQ2MzY1MzczNDk3MjY2NzY2OTcyNDczNTdhMzQ2NzYzNzU0NTJmNjU2ZTY4NjQ0NjZjNmY3ODU4MzE1YTY1Nzg1NTJmMzM0NDcwNzQzOTJmNjI3MTY4NmU1NDRjNjY3NTUyMzM2ZDZhNzU2MjYxNTg0ZjUwMzM1MDM3NzI2ZDZkMzE3NTU4NTk3NTY2MzE2MjZjNGE1NDUwNGQzODMzNDY0MjUwNzgyYjY0NzA2NjU2NTk3NzQ2NDE3OTY2NDI0NzRmNDUyYjUwNjY2MzRiMzU3NTZjNjM2NzZkMzQyZjc1N2E2YjZjNjE1Nzc1NjU2OTY1NmIzODM0NmE1MjJiMzc0ZjQ0NTE1MDRkNTg2ZDYxNmEzOTQ2NGY3OTQ5NTE0NzUxNTkzMjQ2NGM1NjZlNDE2ZTZiNzY1NDcwNTA2ZTY0NDY0ZDU0NzQ2ZDQ5NGYzMDY5Nzk1NDMzNTg0NDdhNTM0NDQ5MzQ3YTQ4Njk1MTUxMzU1MzU0NDU0MTZlNTM0MjMzNTk1NzRlNDg2ZDMxNjg2YjU5NTQ3MjQ0MzM0MTdhNzU0OTM4NDQ3OTRiNTUzODVhNGIzNDc4NzMzNDc5Nzc1NjUyNGQ0ZTZmNzM3YTMwNzA1YTc2NGY2ZDRhMzA0MjRjNzg0NDY5NTk1ODU5NDQ0YjU2Mzg0OTY1Nzg0MjYyNzk2YzQxNzc0MzQ0MmI1MzU2NTQ1MzY5N2E3Nzc1NmEyZjQzNTg2MTU1NDUzODYzNjc0OTRjNTU1OTM4NTk1MDcyNmU2ZDc4NzI0NzRjNjMzNDQ3NzY0MzU1NzM3NzJmNzM0NzQ3NzE0ZTY4NDUzNjc5MzM1NDJmNGU3NzYyNjg2ODRlNTgzMjY2MzQzOTRkNDczMjU2NGY2ZjY3NTU0MjQ5Njc2ZDc2NTY1MDM2NzU3OTUwNDQ0ZjUyNjc0OTU0NTc2NzU4MzQ2NDQxNTMzNTQxNDM0YTcyNzQ0ZTU2NzEzNzdhNzU2YTQ1NTczMTRmNjQ3NDY5NzE1YTY5NTE3ODdhNzc0NzZmNGU3MDQyN2E3MTc3NGM3NzMxNTU0YTU3NTk0YjU4NmI3NjZmMzQ0NDM5NzE3YTQ2NDU1MzQ4NjMzNjQzNTE1MDZiNTUzMjMyNTM0MzM0NWE3MTcyNTY2OTQ4NjI1NTZkNzI2NTQ0NGE0NDYxNmU3ODZmNjc3YTM1NjM2YzVhNjM3MzU2NDM3YTU4NTE2ZjY3NTI1NTQyMzA1MTM0NmY1OTU4MzkzMzc5NTgzMjMzNjM0NzQ1NjMzNTUwNDc2NTRiNTc0NDY0MzAzNDY3NDI2ZDZlMzM2YTc2NTQ3YTUyNzI0NTVhNDQzNzY2NmY1MzcwNGI2YTMzMmY0YzU1NDc2NDc3NDU1NTUyNzM1NDdhNDQzMDY1NDQzOTc2MzczNzY1MmI2YzMxNDU0MjZiNDg0MTY2NTA0YzY1NDI2ZTJiNmQ1OTU2NDYzMDU4NjIzNjUwN2E0NzQxNmM0ZDUxNzM3Mzc3NzI3YTU5NzU0MzMyNDczNTM2NTczNzUyNTU2ZTU1NDczMzUwNzc2ZDM0Nzc0YjRiN2E0OTc4NTIzNzYyMzU3MTY0NjM1MzUxNjI2ZTQ1MzM3MDM2NTM2Nzc1NmI0YzU3NTc1MTM3NGQ1MzMzNzI2YzQxNTA3MTMwNzg3YTQ2NzczODZiNDMzMzZlNzc2NzRkNjU2ZDc1NGU2NTUxNmQ2NTUyMzg0MTZkNjQ3MDQ1NzE0NDM5NGU0NDMyN2E0ODY4NmM0NDU1NDY3MTc5NDI3YTY2NGMzMDMyMzI0OTcxNTg2ZTQxNjY3OTY5MzI1NjMyNjE2NTcyNDM2MzRkNTQ2NzM0NDc1NzM2NmU0NjMzNzg0ODMwN2E2OTZkNGY3OTYyNTI3OTY1NTM1MTY5NTY3MTRjMzg3ODc5NzUzMjJmNTk1MzQ5NTc1NTc4NjY3MzMzNGI3Njc3NDM0NTc5NTY1ODQxM2QzZA=="`;

exports[`should forget the password 3`] = `"eyJpZCI6InV1aWQiLCJjb250YWN0cyI6e30sImRhdGEiOnt9LCJub3RpZmljYXRpb25zIjp7fSwicGVlcnMiOnt9LCJwbHVnaW5zIjp7fSwic2V0dGluZ3MiOnt9LCJ3YWxsZXRzIjp7fX0="`;
